pkg/ring/ring.go

```diff
@@ -59,6 +59,13 @@ type ReadRing interface {
 	// the input operation.
 	GetReplicationSetForOperation(op Operation) (ReplicationSet, error)
 
+	// GetReplicationSetForOperationWithNoQuorum returns all instances where the input operation should be executed.
+	// The resulting ReplicationSet contains all healthy instances in the ring, but the computation for MaxErrors
+	// does not require quorum so only 1 replica is needed to complete the operation. For MaxUnavailableZones, it is
+	// not automatically reduced when there are unhealthy instances in a zone because healthy instances in the zone
+	// are still returned, but the information about zones with unhealthy instances is returned.
+	GetReplicationSetForOperationWithNoQuorum(op Operation) (ReplicationSet, map[string]struct{}, error)
```

Comment:
Nit about the method name. Should we call it `WithoutQuorum` or `WithNoQuorum`?

> For MaxUnavailableZones, it is not automatically reduced when there are unhealthy instances in a zone because healthy instances in the zone are still returned, but the information about zones with unhealthy instances is returned.

It feels like this method differs from `GetReplicationSetForOperation` more than not requiring quorum but also relaxes the constraint where we require all instances in the zone are healthy. 

Why do we make this change? I am afraid that in the future we have a use case where we don't need quorum but we want to exclude zones with unhealthy instances. Then we have to change the interface again.

Reply:
To me, I think it makes sense for the Ruler API backup usecase to return all instances even in an unhealthy zone. However, I feel it is for this usecase only.
Do we think we can remove this method from the interface and move it to Ruler Ring code only? Since it might only make sense for Ruler.

Reply:
Yes I agree, my problem was that there is no ring operation that allows me to get healthy instances as well as unhealthy instances so I can write code in ruler to decide how much error we can tolerate. There is `GetInstanceDescsForOperation` but this only returns healthy instances.  Even if we set the `op` to consider all statuses as healthy it still doesn't guarantee that we get all instances in the ring because `storageLastUpdate := r.KVClient.LastUpdateTime(r.key)` is also considered when checking if the instance is healthy.  Do you think we can add a new method in the ring that could look like `GetAllInstanceDescs(op Operation) ([]InstanceDesc, []InstanceDesc, err)` that returns the healthy and unhealthy instances in the ring?

Reply:
> Do you think we can add a new method in the ring that could look like GetAllInstanceDescs(op Operation) ([]InstanceDesc, []InstanceDesc, err) that returns the healthy and unhealthy instances in the ring

I think we can do that and I prefer this solution. WDYT?

Reply:
We can change `GetInstanceDescsForOperation` as well. WDYT @alanprot

Reply:
Seems like `GetInstanceDescsForOperation` is only used right now in store-gateway sync. This doesn't run very often so i guess an addition of a new slice in the method would be fine.

Reply:
humm...

I think i prefer a new method as well (`GetAllInstanceDescs(op Operation) ([]InstanceDesc, []InstanceDesc, err)`)

Reply:
Updated @yeya24 , PTAL, thank you